wsmaster/che-core-api-devfile/src/test/java/org/eclipse/che/api/devfile/server/DevfileIntegrityValidatorTest.java

```diff
@@ -0,0 +1,153 @@
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp = "Duplicate command name found:'.*'")
+  public void shouldThrowExceptionOnDuplicateCommandName() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getCommands().add(new Command().withName(initialDevfile.getCommands().get(0).getName()));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp =
+          "Found actions which refers to non-existing tools in command '.*':'no_such_tool'")
+  public void shouldThrowExceptionOnUnexistingCommandActionTool() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getCommands().get(0).getActions().add(new Action().withTool("no_such_tool"));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp = "Duplicate project name found:'.*'")
+  public void shouldThrowExceptionOnDuplicateProjectName() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getProjects().add(new Project().withName(initialDevfile.getProjects().get(0).getName()));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp =
+          "Invalid project name found:'.*'. Name must contain only Latin letters,\\n"
+              + "digits or these following special characters ._-")
+  public void shouldThrowExceptionOnInvalidProjectName() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getProjects().get(0).setName("./" + initialDevfile.getProjects().get(0).getName());
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  private Devfile copyOf(Devfile source) {
+    Devfile result = new Devfile();
+    result.setName(source.getName());
+    result.setSpecVersion(source.getSpecVersion());
+    List<Project> projects = new ArrayList<>();
+    for (Project project : source.getProjects()) {
+      projects.add(
+          new Project()
+              .withName(project.getName())
+              .withSource(
+                  new Source()
+                      .withType(project.getSource().getType())
+                      .withLocation(project.getSource().getType())));
+    }
+    result.setProjects(projects);
+    List<Tool> tools = new ArrayList<>();
+    for (Tool tool : source.getTools()) {
+      tools.add(new Tool().withId(tool.getId()).withName(tool.getName()).withType(tool.getType()));
+    }
+    result.setTools(tools);
+    List<Command> commands = new ArrayList<>();
+    for (Command command : source.getCommands()) {
+      List<Action> actions = new ArrayList<>();
+      for (Action action : command.getActions()) {
+        actions.add(
+            new Action()
```

Comment:
I think it would be better to add copy constructor to `Action` and then copy actions list as 
```
commands.stream().map(Action::new).collect(toCollection(ArrayList::new));
```

Reply:
It's autogenerated. Will look for this option in generator (add copying constructors)